word_search.py

```diff
@@ -0,0 +1,102 @@
+'''
+Created on 9 June 2017
+
+@author: Eva
+'''
+from queue import Queue
+
+class WordSearch:
+    
+    def __init__(self):
+        return
+        
+    def words_found_in_grid(self, rowNo, columnNo, grid, dictionary):
+        """Uses bfs to find valid words from a grid according to a given dictionary 
+        
+            Args:
+                rowNo: number of rows in the grid
+                columnNo: number of columns in the grid
+                grid: two-dimensional array of characters
+                dictionary: a dictionary object that holds the valid words and prefices
+            
+            Returns:
+                A list of words that were found on the grid
+        """
+        self.__pathsTaken = []
+        self.__foundWords = []
+        self.__words_to_search = Queue()
+        row = 0
+        column = 0
+        
+        self.__words_to_search.put([grid[column][row], column, row, []])
+        
+        while not self.__words_to_search.empty():
+            currentArray = self.__words_to_search.get()
+            currentWord = currentArray[0]
+            column = currentArray[1]
+            row = currentArray[2]
+            positions = currentArray[3]
+            
+            if dictionary.isPrefix(currentWord):
+                tempPositions = positions.copy()
+                tempPositions.append([column, row])
+                if tempPositions in self.__pathsTaken: continue
```

Comment:
You should not visit the same tempPositions more than once because each time you follow a different path.